File: internal/crypto/seal/seal.go
package seal

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/ecdh"
	"crypto/hpke"
	"crypto/rand"
	"encoding/binary"
	"errors"
	"fmt"

	"github.com/google/uuid"
)

// KeyLen is the size of an X25519 key and of a content key.
const KeyLen = 32

// suite returns the frozen cipher suite. Its components are fixed in code
// rather than configured, so there is nothing to misconfigure and nothing to
// negotiate down.
func suite() (hpke.KEM, hpke.KDF, hpke.AEAD) {
	return hpke.DHKEM(ecdh.X25519()), hpke.HKDFSHA256(), hpke.AES256GCM()
}

// PublicKey is a tenant's archive public key.
type PublicKey struct {
	raw []byte
	pk  hpke.PublicKey
}

// ParsePublicKey deserialises a 32-byte X25519 public key.
func ParsePublicKey(b []byte) (PublicKey, error) {
	if len(b) != KeyLen {
		return PublicKey{}, fmt.Errorf("seal: public key must be %d bytes, got %d", KeyLen, len(b))
	}
	kem, _, _ := suite()
	pk, err := kem.NewPublicKey(b)
	if err != nil {
		return PublicKey{}, fmt.Errorf("seal: invalid public key: %w", err)
	}
	return PublicKey{raw: append([]byte(nil), b...), pk: pk}, nil
}

// Bytes returns the serialised key.
func (p PublicKey) Bytes() []byte { return append([]byte(nil), p.raw...) }

// Valid reports whether the key was parsed.
func (p PublicKey) Valid() bool { return p.pk != nil }

// PrivateKey is a tenant's archive private key.
//
// It exists in this package so that tests, the CLI and the eventual client
// tooling can open what the server sealed. The server itself never constructs
// one: nothing in the ingest path can, because nothing gives it the bytes.
type PrivateKey struct {
	sk hpke.PrivateKey
}

// GenerateKeyPair creates a tenant archive keypair.
//
// In production this runs in the founder's browser and only the public half is
// ever transmitted. The server-side implementation exists for tests and for
// tooling that runs on an operator's own machine.
func GenerateKeyPair() (PublicKey, PrivateKey, error) {
	kem, _, _ := suite()
	sk, err := kem.GenerateKey()
	if err != nil {
		return PublicKey{}, PrivateKey{}, fmt.Errorf("seal: generate key: %w", err)
	}
	pub, err := ParsePublicKey(sk.PublicKey().Bytes())
	if err != nil {
		return PublicKey{}, PrivateKey{}, err
	}
	return pub, PrivateKey{sk: sk}, nil
}

// ParsePrivateKey deserialises a private key.
func ParsePrivateKey(b []byte) (PrivateKey, error) {
	if len(b) != KeyLen {
		return PrivateKey{}, fmt.Errorf("seal: private key must be %d bytes, got %d", KeyLen, len(b))
	}
	kem, _, _ := suite()
	sk, err := kem.NewPrivateKey(b)
	if err != nil {
		return PrivateKey{}, fmt.Errorf("seal: invalid private key: %w", err)
	}
	return PrivateKey{sk: sk}, nil
}

// Bytes serialises the private key.
func (p PrivateKey) Bytes() ([]byte, error) {
	if p.sk == nil {
		return nil, errors.New("seal: no private key")
	}
	return p.sk.Bytes()
}

// PublicKey returns the matching public key.
func (p PrivateKey) PublicKey() (PublicKey, error) {
	if p.sk == nil {
		return PublicKey{}, errors.New("seal: no private key")
	}
	return ParsePublicKey(p.sk.PublicKey().Bytes())
}

// Valid reports whether the key was parsed.
func (p PrivateKey) Valid() bool { return p.sk != nil }

// ---------------------------------------------------------------------------
// Direct mode: sealed straight to the tenant key.
// ---------------------------------------------------------------------------

// SealDirect seals a small, low-volume payload to the tenant public key.
//
// Used for content keys and key grants. Never for message content: one
// asymmetric operation per message is what makes a phone take half a minute to
// open a conversation.
func SealDirect(pub PublicKey, kind Kind, tenant, row uuid.UUID, epoch uint16, plaintext []byte) ([]byte, error) {
	if !pub.Valid() {
		return nil, errors.New("seal: no public key")
	}
	_, kdf, aead := suite()
	hdr := header{Version: Version, Suite: SuiteV1, Mode: ModeDirect, Epoch: epoch}.encode()

	enc, sender, err := hpke.NewSender(pub.pk, kdf, aead, infoFor(kind, tenant, epoch))
	if err != nil {
		return nil, fmt.Errorf("seal: hpke sender: %w", err)
	}
	ct, err := sender.Seal(buildAAD(kind, tenant, row, hdr), plaintext)
	if err != nil {
		return nil, fmt.Errorf("seal: hpke seal: %w", err)
	}

	out := make([]byte, 0, headerLen+len(enc)+len(ct))
	out = append(out, hdr...)
	out = append(out, enc...)
	return append(out, ct...), nil
}

// OpenDirect reverses SealDirect. Client side only.
func OpenDirect(priv PrivateKey, kind Kind, tenant, row uuid.UUID, envelope []byte) ([]byte, error) {
	if !priv.Valid() {
		return nil, errors.New("seal: no private key")
	}
	h, err := parseHeader(envelope)
	if err != nil {
		return nil, err
	}
	if h.Mode != ModeDirect {
		return nil, fmt.Errorf("%w: expected direct, got %#x", ErrMode, h.Mode)
	}
	if len(envelope) < headerLen+encLen+tagLen {
		return nil, ErrShort
	}

	enc := envelope[headerLen : headerLen+encLen]
	ct := envelope[headerLen+encLen:]

	_, kdf, aead := suite()
	recipient, err := hpke.NewRecipient(enc, priv.sk, kdf, aead, infoFor(kind, tenant, h.Epoch))
	if err != nil {
		return nil, fmt.Errorf("seal: hpke recipient: %w", err)
	}
	pt, err := recipient.Open(buildAAD(kind, tenant, row, envelope[:headerLen]), ct)
	if err != nil {
		// Deliberately opaque. Distinguishing "wrong key" from "tampered" from
		// "wrong row" would hand an attacker with database write access an
		// oracle for probing the binding.
		return nil, ErrAuthentication
	}
	return pt, nil
}

// ---------------------------------------------------------------------------
// Batch mode: sealed under a content key.
// ---------------------------------------------------------------------------

// ContentKey is a symmetric key covering a batch of sealed values.
type ContentKey struct {
	// ID is a per-tenant counter, carried in the envelope so a reader knows
	// which key to unwrap. Scoped per tenant so it does not leak system-wide
	// volume to clients.
	ID    uint32
	Epoch uint16

	key []byte // 32 bytes, AES-256-GCM

	// Sealed is the key itself sealed to the tenant public key. This is what
	// is stored; the raw key never touches disk.
	Sealed []byte
}

// NewContentKey generates a content key and seals it to one device's archive key.
//
// Per device rather than per tenant. With one key for a whole tenant, "this
// operator may read that WhatsApp account and not this one" is a rule the
// server enforces — and the premise here is that server-enforced rules do not
// survive a compromised server. Sealing to the device makes the separation
// arithmetic: a reader holding one device's key cannot open another's.
func NewContentKey(pub PublicKey, tenant, device uuid.UUID, epoch uint16, id uint32) (*ContentKey, error) {
	key := make([]byte, KeyLen)
	if _, err := rand.Read(key); err != nil {
		return nil, fmt.Errorf("seal: entropy: %w", err)
	}
	// The row identity is derived from the device and the id, so a key sealed
	// for one slot cannot be presented as another — or as another device's.
	sealed, err := SealDirect(pub, KindContentKey, tenant, ContentKeyRow(tenant, device, id), epoch, key)
	if err != nil {
		return nil, err
	}
	return &ContentKey{ID: id, Epoch: epoch, key: key, Sealed: sealed}, nil
}

// OpenContentKey unwraps a stored content key. Client side only.
func OpenContentKey(priv PrivateKey, tenant, device uuid.UUID, id uint32, sealed []byte) (*ContentKey, error) {
	h, err := parseHeader(sealed)
	if err != nil {
		return nil, err
	}
	key, err := OpenDirect(priv, KindContentKey, tenant, ContentKeyRow(tenant, device, id), sealed)
	if err != nil {
		return nil, err
	}
	if len(key) != KeyLen {
		return nil, fmt.Errorf("seal: content key is %d bytes, want %d", len(key), KeyLen)
	}
	return &ContentKey{ID: id, Epoch: h.Epoch, key: key, Sealed: sealed}, nil
}

// ContentKeyRow derives the stable uuid a content key binds to.
//
// Exported because the browser has to compute the same value to open one, and
// reimplementing a derivation from prose is how two sides quietly disagree.
func ContentKeyRow(tenant, device uuid.UUID, id uint32) uuid.UUID {
	name := make([]byte, 0, 16+4)
	name = append(name, device[:]...)
	name = binary.BigEndian.AppendUint32(name, id)
	return uuid.NewSHA1(tenant, name)
}

// GrantRow derives the uuid a key grant binds to.
//
// A grant is one device's private archive key sealed to one person's public
// key. Binding it to (device, user, epoch) is what stops a grant issued for one
// device being presented as the grant for another.
func GrantRow(tenant, device, user uuid.UUID, epoch uint16) uuid.UUID {
	name := make([]byte, 0, 16+16+2)
	name = append(name, device[:]...)
	name = append(name, user[:]...)
	name = binary.BigEndian.AppendUint16(name, epoch)
	return uuid.NewSHA1(tenant, name)
}

// Seal encrypts a value under this content key.
func (c *ContentKey) Seal(kind Kind, tenant, row uuid.UUID, plaintext []byte) ([]byte, error) {
	gcm, err := c.gcm()
	if err != nil {
		return nil, err
	}
	nonce := make([]byte, nonceLen)
	if _, err := rand.Read(nonce); err != nil {
		return nil, fmt.Errorf("seal: entropy: %w", err)
	}

	hdr := header{Version: Version, Suite: SuiteV1, Mode: ModeBatch, Epoch: c.Epoch}.encode()

	out := make([]byte, 0, BatchOverhead+len(plaintext))
	out = append(out, hdr...)
	out = binary.BigEndian.AppendUint32(out, c.ID)
	out = append(out, nonce...)
	return gcm.Seal(out, nonce, plaintext, buildAAD(kind, tenant, row, hdr)), nil
}

// Open decrypts a value sealed under this content key. Client side only.
func (c *ContentKey) Open(kind Kind, tenant, row uuid.UUID, envelope []byte) ([]byte, error) {
	h, err := parseHeader(envelope)
	if err != nil {
		return nil, err
	}
	if h.Mode != ModeBatch {
		return nil, fmt.Errorf("%w: expected batch, got %#x", ErrMode, h.Mode)
	}
	if len(envelope) < BatchOverhead {
		return nil, ErrShort
	}
	if id := binary.BigEndian.Uint32(envelope[headerLen : headerLen+4]); id != c.ID {
		return nil, fmt.Errorf("seal: envelope needs content key %d, have %d", id, c.ID)
	}

	gcm, err := c.gcm()
	if err != nil {
		return nil, err
	}
	nonce := envelope[headerLen+4 : headerLen+4+nonceLen]
	ct := envelope[headerLen+4+nonceLen:]
	pt, err := gcm.Open(nil, nonce, ct, buildAAD(kind, tenant, row, envelope[:headerLen]))
	if err != nil {
		return nil, ErrAuthentication
	}
	return pt, nil
}

func (c *ContentKey) gcm() (cipher.AEAD, error) {
	block, err := aes.NewCipher(c.key)
	if err != nil {
		return nil, fmt.Errorf("seal: aes: %w", err)
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, fmt.Errorf("seal: gcm: %w", err)
	}
	return gcm, nil
}

// ContentKeyID reads which content key an envelope needs, without opening it.
// The dispatcher uses this to fetch the right key.
func ContentKeyID(envelope []byte) (id uint32, epoch uint16, err error) {
	h, err := parseHeader(envelope)
	if err != nil {
		return 0, 0, err
	}
	if h.Mode != ModeBatch {
		return 0, h.Epoch, fmt.Errorf("%w: not a batch envelope", ErrMode)
	}
	if len(envelope) < headerLen+4 {
		return 0, 0, ErrShort
	}
	return binary.BigEndian.Uint32(envelope[headerLen : headerLen+4]), h.Epoch, nil
}

File: web/src/crypto/seal.ts
// The sealed archive.
//
// Mirrors internal/crypto/seal. The server holds only a public key; everything
// below runs in the browser with the private half, which is the only place it
// ever exists.
//
// Pinned against Go by test/seal.spec.ts, which opens the vectors that package
// generates — including the ones that must fail.

import {
  type Bytes,
  concat,
  encodeUTF8,
  formatUUID,
  readUint16BE,
  readUint32BE,
  uuidV5,
} from './bytes'
import { ENC_LEN, open as hpkeOpen, seal as hpkeSeal, type PrivateKey } from './hpke'

export const MAGIC0 = 0x57 // 'W'
export const MAGIC1 = 0x53 // 'S'
export const VERSION = 0x01
export const SUITE_V1 = 0x01

export const MODE_DIRECT = 0x01
export const MODE_BATCH = 0x02

const HEADER_LEN = 8
const NONCE_LEN = 12
const TAG_LEN = 16

/** Kind identifies what a sealed value is, and is bound into the AAD. */
export enum Kind {
  Body = 0x01,
  RawProto = 0x02,
  MediaKey = 0x03,
  Thumbnail = 0x04,
  ContactName = 0x05,
  ContentKey = 0x06,
  DeviceGrant = 0x07,
  UserWrap = 0x08,
  Payload = 0x09,
  PushName = 0x0a,
  FullName = 0x0b,
  BusinessName = 0x0c,
  Avatar = 0x0d,
}

// The names go into the HPKE info string, so they are part of the wire format
// rather than labels for humans. They must match Kind.String() in Go exactly.
const kindNames: Record<number, string> = {
  [Kind.Body]: 'body',
  [Kind.RawProto]: 'raw_proto',
  [Kind.MediaKey]: 'media_key',
  [Kind.Thumbnail]: 'thumbnail',
  [Kind.ContactName]: 'contact_name',
  [Kind.ContentKey]: 'content_key',
  [Kind.DeviceGrant]: 'device_grant',
  [Kind.UserWrap]: 'user_wrap',
  [Kind.Payload]: 'payload',
  [Kind.PushName]: 'push_name',
  [Kind.FullName]: 'full_name',
  [Kind.BusinessName]: 'business_name',
  [Kind.Avatar]: 'avatar',
}

export function kindName(kind: number): string {
  // Go's default branch is fmt.Sprintf("kind(%#x)", byte(k)), which prints the
  // hex without leading zeros.
  return kindNames[kind] ?? `kind(0x${kind.toString(16)})`
}

/** SealError distinguishes a wrong key from a blob that was moved or altered. */
export class SealError extends Error {
  constructor(
    message: string,
    readonly code:
      | 'short'
      | 'magic'
      | 'version'
      | 'suite'
      | 'mode'
      | 'key_mismatch'
      | 'authentication',
  ) {
    super(message)
    this.name = 'SealError'
  }
}

export interface Header {
  version: number
  suite: number
  mode: number
  epoch: number
}

/**
 * parseHeader validates the prefix before any key material is touched.
 *
 * The suite byte is checked here rather than guessed at, because a parser that
 * guessed would be the downgrade path the byte exists to close.
 */
export function parseHeader(envelope: Bytes): Header {
  if (envelope.length < HEADER_LEN) throw new SealError('envelope curto demais', 'short')
  if (envelope[0] !== MAGIC0 || envelope[1] !== MAGIC1) {
    throw new SealError('não é um envelope', 'magic')
  }
  const h: Header = {
    version: envelope[2],
    suite: envelope[3],
    mode: envelope[4],
    epoch: readUint16BE(envelope, 5),
  }
  if (h.version !== VERSION) throw new SealError(`versão desconhecida: ${h.version}`, 'version')
  if (h.suite !== SUITE_V1) throw new SealError(`suíte desconhecida: ${h.suite}`, 'suite')
  if (h.mode !== MODE_DIRECT && h.mode !== MODE_BATCH) {
    throw new SealError(`modo desconhecido: ${h.mode}`, 'mode')
  }
  return h
}

const AAD_PREFIX = encodeUTF8('wsv1')

/**
 * buildAAD binds a sealed value to where it lives and to its own header.
 *
 * Leaving any of these out is the failure mode that matters: an attacker with
 * write access to the database moves a blob to another row, and a client that
 * did not bind the row decrypts it happily — content swapped between
 * conversations, with no integrity failure anywhere.
 */
function buildAAD(kind: Kind, tenant: Bytes, row: Bytes, header: Bytes): Bytes {
  return concat(AAD_PREFIX, new Uint8Array([kind]), tenant, row, header)
}

/** infoFor gives each use of the tenant key its own HPKE domain. */
function infoFor(kind: Kind, tenant: Bytes, epoch: number): Bytes {
  return encodeUTF8(`wsv1/${kindName(kind)}/${formatUUID(tenant)}/${epoch}`)
}

/**
 * encodeHeader writes the eight bytes every envelope starts with.
 *
 * The last byte is reserved and must be zero: it is part of the AAD, so a
 * client that wrote anything else there would produce blobs the server's own
 * reader rejects.
 */
function encodeHeader(mode: number, epoch: number): Bytes {
  const b = new Uint8Array(HEADER_LEN) as Bytes
  b[0] = MAGIC0
  b[1] = MAGIC1
  b[2] = VERSION
  b[3] = SUITE_V1
  b[4] = mode
  new DataView(b.buffer).setUint16(5, epoch, false)
  return b
}

/**
 * sealDirect seals a value straight to a public key.
 *
 * The only thing this client seals is a key grant: one device's private archive
 * key, encrypted to an account's public key so that person can read that
 * WhatsApp number by signing in. It happens here because the device key is
 * generated here and must not leave — the server stores ciphertext it cannot
 * open, which is the whole arrangement.
 *
 * Getting the binding wrong fails silently in the worst possible way: the grant
 * stores, pairing reports success, and the archive it was supposed to unlock is
 * unreadable by anyone, forever. test/grant.spec.ts seals a vector here that Go
 * has to open.
 */
export async function sealDirect(
  publicRaw: Bytes,
  kind: Kind,
  tenant: Bytes,
  row: Bytes,
  epoch: number,
  plaintext: Bytes,
): Promise<Bytes> {
  const header = encodeHeader(MODE_DIRECT, epoch)
  const { enc, ciphertext } = await hpkeSeal(
    publicRaw,
    infoFor(kind, tenant, epoch),
    buildAAD(kind, tenant, row, header),
    plaintext,
  )
  return concat(header, enc, ciphertext)
}

/**
 * openDirect opens a value sealed straight to the archive key.
 *
 * Used only for content keys and grants. One asymmetric operation per value is
 * what makes opening a large history slow, which is why message content never
 * takes this path.
 */
export async function openDirect(
  priv: PrivateKey,
  kind: Kind,
  tenant: Bytes,
  row: Bytes,
  envelope: Bytes,
): Promise<Bytes> {
  const h = parseHeader(envelope)
  if (h.mode !== MODE_DIRECT) throw new SealError('esperava modo direto', 'mode')
  if (envelope.length < HEADER_LEN + ENC_LEN + TAG_LEN) {
    throw new SealError('envelope curto demais', 'short')
  }

  const header = envelope.subarray(0, HEADER_LEN)
  const enc = envelope.subarray(HEADER_LEN, HEADER_LEN + ENC_LEN)
  const ciphertext = envelope.subarray(HEADER_LEN + ENC_LEN)

  try {
    return await hpkeOpen(
      priv,
      enc,
      infoFor(kind, tenant, h.epoch),
      buildAAD(kind, tenant, row, header),
      ciphertext,
    )
  } catch {
    // Deliberately opaque, as on the server: telling "wrong key" apart from
    // "moved row" apart from "tampered" hands an attacker an oracle for
    // probing the binding.
    throw new SealError('autenticação falhou', 'authentication')
  }
}

/**
 * ContentKey is a symmetric key covering a batch of sealed values.
 *
 * Batching is a requirement, not a tuning knob: one X25519 operation per
 * message is twelve to thirty seconds to open a ten-thousand message history on
 * a mid-range phone, and it recurs on every new device.
 */
export class ContentKey {
  private constructor(
    readonly id: number,
    readonly epoch: number,
    private readonly key: CryptoKey,
  ) {}

  /**
   * unwrap opens a stored content key with a device's archive private key.
   *
   * The device is part of it because the archive key is per device: content
   * key ids are counted per device, so id 7 names a different key on every
   * account, and the row the value binds to is derived from both.
   */
  static async unwrap(
    priv: PrivateKey,
    tenant: Bytes,
    device: Bytes,
    id: number,
    sealed: Bytes,
  ): Promise<ContentKey> {
    const h = parseHeader(sealed)
    const row = await contentKeyRow(tenant, device, id)
    const raw = await openDirect(priv, Kind.ContentKey, tenant, row, sealed)
    if (raw.length !== 32) {
      throw new SealError(`a chave de conteúdo tem ${raw.length} bytes, esperava 32`, 'short')
    }
    const key = await crypto.subtle.importKey('raw', raw, { name: 'AES-GCM' }, false, ['decrypt'])
    return new ContentKey(id, h.epoch, key)
  }

  async open(
    kind: Kind,
    tenant: Bytes,
    row: Bytes,
    envelope: Bytes,
  ): Promise<Bytes> {
    const h = parseHeader(envelope)
    if (h.mode !== MODE_BATCH) throw new SealError('esperava modo em lote', 'mode')
    if (envelope.length < HEADER_LEN + 4 + NONCE_LEN + TAG_LEN) {
      throw new SealError('envelope curto demais', 'short')
    }
    const wants = readUint32BE(envelope, HEADER_LEN)
    if (wants !== this.id) {
      throw new SealError(
        `o envelope precisa da chave de conteúdo ${wants}, esta é a ${this.id}`,
        'key_mismatch',
      )
    }

    const header = envelope.subarray(0, HEADER_LEN)
    const nonce = envelope.subarray(HEADER_LEN + 4, HEADER_LEN + 4 + NONCE_LEN)
    const ciphertext = envelope.subarray(HEADER_LEN + 4 + NONCE_LEN)

    try {
      const plaintext = await crypto.subtle.decrypt(
        {
          name: 'AES-GCM',
          iv: nonce,
          additionalData: buildAAD(kind, tenant, row, header),
          tagLength: 128,
        },
        this.key,
        ciphertext,
      )
      return new Uint8Array(plaintext)
    } catch {
      throw new SealError('autenticação falhou', 'authentication')
    }
  }
}

/**
 * contentKeyRow derives the row identity a content key is bound to.
 *
 * Derived rather than stored, so a key sealed for one slot cannot be presented
 * as another — or as another device's.
 */
async function contentKeyRow(tenant: Bytes, device: Bytes, id: number): Promise<Bytes> {
  const name = new Uint8Array(20)
  name.set(device, 0)
  new DataView(name.buffer).setUint32(16, id, false)
  return uuidV5(tenant, name)
}

/**
 * grantRow derives the row a key grant binds to.
 *
 * A grant is one device's private archive key sealed to one person's public
 * key. Binding it to (device, user, epoch) stops a grant issued for one device
 * being presented as the grant for another.
 */
export async function grantRow(
  tenant: Bytes,
  device: Bytes,
  user: Bytes,
  epoch: number,
): Promise<Bytes> {
  const name = new Uint8Array(34)
  name.set(device, 0)
  name.set(user, 16)
  new DataView(name.buffer).setUint16(32, epoch, false)
  return uuidV5(tenant, name)
}
